[TOPIC-BLE-LLCP]: Bluetooth: llcp: release ctx at end of procedure

kruithofa opened the pull request:
When a control procedure is finished, or upon disconnection, we need
to release the context to avoid running out of memory.
I believe I found the best location for this, pls .correct me if I am wrong

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

Commit:
[TOPIC-BLE-LLCP]: Bluetooth: llcp: release ctx at end of procedure

When a control procedure is finished, or upon disconnection, we need
to release the context to avoid running out of memory

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

thoh-ot commented:
Would say a better approach would be to mark the for release and then in the end of run/rx check for said release and do so, this way you prevent any weird access to the ctx on the return path of the call stack.

For disconnect you need to empty the lists completely to remove all pending procedures (do we need to generate any host notification here???)

kruithofa commented:
I'll look into your suggestion

kruithofa commented:
We may need to keep an eye on the behaviour of the phy-update procedure when doing the integration with LL, in particular when receiving a REJECT_IND. Unit tests pass, but the act_run procedure gets called when the control procedure is finished which IMO is unexpected.

Commit:
[TOPIC-BLE-LLCP]: Bluetooth: llcp: unittest for context release

Adds unit tests for the context release when disconnected in the
middle of a control procedure

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>
